mixin: Add max_samples_per_send to the remote write dashboard

jorgelbg opened the pull request:
Adds a graph panel with the `max_samples_per_send` metric (introduced in PR #8102) to the Shard Details row of the remote write dashboard.

Signed-off-by: Jorge Luis Betancourt Gonzalez <jorge-luis.betancourt@trivago.com>

/cc @csmarchbanks @cstyan

jorgelbg commented:
@cstyan I see two main motivations when a user needs to look at this metric:

* One: Purely informational to see/know what value is configured/used.
* Two: When troubleshooting the `remote_write` config because sample ingestion is lagging.

For the first scenario, a stat panel is perfect (even more with the repeat option). The graph panel is not particularly useful here. If all is working fine most of the time it will plot the same value (per remote write config).

For the second scenario, I found that a normal graph panel is usually better because it's quite likely that the user needs to correlate the number of dropped/pending samples (and other metrics) with the `max_samples_per_send`. This was what prompted the initial PR. In this case, a stats panel is less useful because we cannot immediately see the previous values, just the overall "trend" with the sparkline view.

I am fine with either option, just wanted to put my view on this.

> If you have the time it might be worth seeing what it would be like to have all the config/shard metric values as stat panels with some kind of repeat, or just converting all the shards metrics to one big table panel where each row is the remote write endpoint.

Since I have some time I'm going to give it a try to transforming the Shard/Shard Details rows to stat panels with the repeat option per remote write URL.

cstyan commented:
`In this case, a stats panel is less useful because we cannot immediately see the previous values, just the overall "trend" with the sparkline view.` Not sure I entirely understand why having `max_samples_per_send` in the graph is much more useful in this case, since changing that value (or any other queue config value) results in a restart of remote write, which would result in a reset of metrics like dropped samples/pending samples. Though I am open to seeing an example of the graph and being proved otherwise :) I think if we were to go that route it would again be best to have a repeat per remote write endpoint so that one config with a much higher/lower config value doesn't make the graph lines for the other endpoints hard to view.

beorn7 commented:
Generally, I tend to not setup dashboards for anticipated troubleshooting scenarios. Dashboards are fine to get an immediate impression of the current load and health situation, and they should give you entry points to drill down (e.g. in explore mode). 

Since `prometheus_remote_storage_max_samples_per_send` just reflects a config setting, I would generally shy away from graphing it. It takes away valuable screen real estate for something mostly boring. A stats panel makes more sense in most situations, as said above, and if you need to drill down because you suspect recent changes of the config, you can start an exploration from that panel.

Having said that, I'm not a remote-write expert, and this specific case might be an exception to my general best practices.

jorgelbg commented:
I migrated the panels in the Shards row to single stat panels. I also included a couple of changes that I leave for your consideration and I'm also happy to revert/change as needed:

* I renamed the Shards row to `Queue [$url]` since (I think) this reflects now that the values shown are tied to a specific remote config queue.
* I also moved the Pending Samples panel to the Sample row, since this is not a configuration option and its value changes over time.
* I also played a bit with the `height` of the single stat panels, it felt like it was using too much vertical space (but I may be wrong on this one).

Here is how the dashboards look like: https://snapshot.raintank.io/dashboard/snapshot/c3oX67oNwMV2BYZsoR3xESjb6v2xbsA3?orgId=2

For now, I kept the position of the Queue row but maybe we can find a better position within the dashboard.

beorn7 commented:
Looks very nice to me at first glance. I leave it to @cstyan to apply more scrutiny. (o:

cstyan commented on documentation/prometheus-mixin/dashboards.libsonnet:
```diff
@@ -148,76 +148,102 @@ local template = grafana.template;
+      local maxSamplesPerSend =
         graphPanel.new(
-          'Current Shards',
+          'Max samples sent per request',
```
can you name this `Max samples per send` so that it matches the queue config fields name.

cstyan commented on documentation/prometheus-mixin/dashboards.libsonnet:
```diff
@@ -148,76 +148,102 @@ local template = grafana.template;
+          'prometheus_remote_storage_shard_capacity{cluster=~"$cluster", instance=~"$instance", url=~"$url"}',
         ));
 
+      local maxSamplesPerSend =
```
this seems to be a duplicate of `maxSamplesPerSend` above, but one is a graph panel and the other is a single stat

cstyan commented on documentation/prometheus-mixin/dashboards.libsonnet:
```diff
@@ -353,20 +379,19 @@ local template = grafana.template;
         row.new(
-          'Shards'
+          'Queue [$url]',
+          repeat='url'
```
There could be multiple remote write configs for the same URL, we probably want to repeat on `remote_name:url`, or maybe also include instance. I don't know if we'll be able to get something that works 100% of the time because of HA setups but it's worth trying some other label combinations here.

cstyan replied:
Hmmm after some quick searching it sounds like repeating on multiple vars is not possible. This may end up being best implemented as a table panel.

roidelapluie replied:
It should be possible with the new Stat panel.

jorgelbg replied:
@roidelapluie you mean repeating on multiple variables or repeating the panel itself? We could repeat the panel (on `remote_name`) inside the row but that leads to each row having multiple "empty" panels for any `url`/`remote_name` pair that does not match.

It ends up looking like this (wasting a lot of space):
![image](https://user-images.githubusercontent.com/1291846/119793412-da51ba00-bed6-11eb-8b9e-e1a31750c853.png)

Unfortunately, hiding empty panels is currently an open issue in Grafana https://github.com/grafana/grafana/issues/4470.

roidelapluie replied:
No, I mean using the Stat panel, which supports multiple values.

roidelapluie replied:
That panel: https://grafana.com/docs/grafana/latest/panels/visualizations/stat-panel/

jorgelbg replied:
Even with the new stats panel I only see the option to repeat using a single variable:

![screenshot of the repeat options of the new Grafana stats panels](https://user-images.githubusercontent.com/1291846/122745245-17685c80-d289-11eb-8401-af22ec74d2b1.png)

I don't see any of selecting multiple variables on that dropdown.

cstyan replied:
Yeah I don't think it's possible to repeat on more than one label. Table is probably the easiest option unless you want to ask around in some of the Grafana communities discussion spots.

beorn7 commented:
We looked at this during our bug scrub. Since the discussion got stuck so long ago, we decided to close this for now. But if anyone believes this should be finalized and merged, feel free to follow-up or re-open.
